# Projeto Ciência de Dados com Python
Abordagem de estudo de caso para a criação de projetos de ciência de dados usando Python, pandas e scikit-learn.
Stephen Klosterman

In [5]:
# importar a biblioteca pandas no env
import pandas as pd

In [6]:
# utilizar o pandas para carregar o arquivo de dados (.xls)
pd.read_excel

<function pandas.io.excel._base.read_excel(io, sheet_name: 'str | int | list[IntStrT] | None' = 0, header: 'int | Sequence[int] | None' = 0, names=None, index_col: 'int | Sequence[int] | None' = None, usecols=None, squeeze: 'bool | None' = None, dtype: 'DtypeArg | None' = None, engine: "Literal['xlrd', 'openpyxl', 'odf', 'pyxlsb'] | None" = None, converters=None, true_values: 'Iterable[Hashable] | None' = None, false_values: 'Iterable[Hashable] | None' = None, skiprows: 'Sequence[int] | int | Callable[[int], object] | None' = None, nrows: 'int | None' = None, na_values=None, keep_default_na: 'bool' = True, na_filter: 'bool' = True, verbose: 'bool' = False, parse_dates=False, date_parser=None, thousands: 'str | None' = None, decimal: 'str' = '.', comment: 'str | None' = None, skipfooter: 'int' = 0, convert_float: 'bool | None' = None, mangle_dupe_cols: 'bool' = True, storage_options: 'StorageOptions' = None) -> 'DataFrame | dict[IntStrT, DataFrame]'>

In [4]:
# supports old-style Excel files(.xls)
pip install xlrd


SyntaxError: invalid syntax (585056676.py, line 2)

In [10]:
# supports newer Excel file formats
pip install openpyxl

  Using cached openpyxl-3.0.9-py2.py3-none-any.whl (242 kB)
  Using cached et_xmlfile-1.1.0-py3-none-any.whl (4.7 kB)
Note: you may need to restart the kernel to use updated packages.


In [7]:
# importar o dataset que está no formato do Excel, como um DataFrame
df = pd.read_excel(r'C:\Users\felip\Desktop\Workflow\Ciência de Dados\Projetos de Ciência de Dados com Python\default_of_credit_card_clients__courseware_version_1_21_19.xls')


In [8]:
# projeta o DataFrame
print(df)

                  ID  LIMIT_BAL  SEX  EDUCATION  MARRIAGE  AGE PAY_1  PAY_2  \
0      798fc410-45c1      20000    2          2         1   24     2      2   
1      8a8c8f3b-8eb4     120000    2          2         2   26    -1      2   
2      85698822-43f5      90000    2          2         2   34     0      0   
3      0737c11b-be42      50000    2          2         1   37     0      0   
4      3b7f77cc-dbc0      50000    1          2         1   57    -1      0   
...              ...        ...  ...        ...       ...  ...   ...    ...   
29995  ecff42d0-bdc6     220000    1          3         1   39     0      0   
29996  99d1fa0e-222b     150000    1          3         2   43    -1     -1   
29997  95cdd3e7-4f24      30000    1          2         2   37     4      3   
29998  00d03f02-04cd      80000    1          3         1   41     1     -1   
29999  15d69f9f-5ad3      50000    1          2         1   46     0      0   

       PAY_3  PAY_4  ...  BILL_AMT4  BILL_AMT5  BIL

In [9]:
# verifiar se o número de linhas e colunas coincidem com o arquivo do Excel
df.shape

(30000, 25)

In [10]:
# fornece a informação sobre todas as colunas
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 30000 entries, 0 to 29999
Data columns (total 25 columns):
 #   Column                      Non-Null Count  Dtype 
---  ------                      --------------  ----- 
 0   ID                          30000 non-null  object
 1   LIMIT_BAL                   30000 non-null  int64 
 2   SEX                         30000 non-null  int64 
 3   EDUCATION                   30000 non-null  int64 
 4   MARRIAGE                    30000 non-null  int64 
 5   AGE                         30000 non-null  int64 
 6   PAY_1                       30000 non-null  object
 7   PAY_2                       30000 non-null  int64 
 8   PAY_3                       30000 non-null  int64 
 9   PAY_4                       30000 non-null  int64 
 10  PAY_5                       30000 non-null  int64 
 11  PAY_6                       30000 non-null  int64 
 12  BILL_AMT1                   30000 non-null  int64 
 13  BILL_AMT2                   30000 non-null  in

In [11]:
# fornece as primeiras linhas do DataFrame
df.head()

ID  LIMIT_BAL  SEX  EDUCATION  MARRIAGE  AGE PAY_1  PAY_2  \
0  798fc410-45c1      20000    2          2         1   24     2      2   
1  8a8c8f3b-8eb4     120000    2          2         2   26    -1      2   
2  85698822-43f5      90000    2          2         2   34     0      0   
3  0737c11b-be42      50000    2          2         1   37     0      0   
4  3b7f77cc-dbc0      50000    1          2         1   57    -1      0   

   PAY_3  PAY_4  ...  BILL_AMT4  BILL_AMT5  BILL_AMT6  PAY_AMT1  PAY_AMT2  \
0     -1     -1  ...          0          0          0         0       689   
1      0      0  ...       3272       3455       3261         0      1000   
2      0      0  ...      14331      14948      15549      1518      1500   
3      0      0  ...      28314      28959      29547      2000      2019   
4     -1      0  ...      20940      19146      19131      2000     36681   

   PAY_AMT3  PAY_AMT4  PAY_AMT5  PAY_AMT6  default payment next month  
0         0         0         0         0                           1  
1      1000      1000         0      2000                           1  
2      1000      1000      1000      5000                           0  
3      1200      1100      1069      1000                           0  
4     10000      9000       689       679                           0  

[5 rows x 25 columns]

In [12]:
# fornece as primeiras n(20) linhas do DataFrame
df.head(20)

ID  LIMIT_BAL  SEX  EDUCATION  MARRIAGE  AGE          PAY_1  \
0   798fc410-45c1      20000    2          2         1   24              2   
1   8a8c8f3b-8eb4     120000    2          2         2   26             -1   
2   85698822-43f5      90000    2          2         2   34              0   
3   0737c11b-be42      50000    2          2         1   37              0   
4   3b7f77cc-dbc0      50000    1          2         1   57             -1   
5   ceae2946-47e4      50000    1          1         2   37              0   
6   d3b8ce26-9c89     500000    1          1         2   29              0   
7   d6e1911b-54d0     100000    2          2         2   23              0   
8   8421d3b3-2b54     140000    2          3         1   28              0   
9   62a2fd79-cd90      20000    1          3         2   35             -2   
10  b73c34fc-91cf     200000    2          3         2   34              0   
11  4777689a-5131     260000    2          1         2   51             -1   
12  b814ac5c-3d88     630000    2          2         2   41             -1   
13  a17b23d4-357b      70000    1          2         2   30              1   
14  0d6f99ad-84e9     250000    1          1         2   29              0   
15  8f023df0-a9ea      50000    2          3         3   23              1   
16  cbf648a6-98aa      20000    1          1         2   24              0   
17  f0ba5e02-0c0f     320000    1          1         1   49  Not available   
18  cb258690-8ea2     360000    2          1         1   49              1   
19  f6f0307a-b75f     180000    2          1         2   29              1   

    PAY_2  PAY_3  PAY_4  ...  BILL_AMT4  BILL_AMT5  BILL_AMT6  PAY_AMT1  \
0       2     -1     -1  ...          0          0          0         0   
1       2      0      0  ...       3272       3455       3261         0   
2       0      0      0  ...      14331      14948      15549      1518   
3       0      0      0  ...      28314      28959      29547      2000   
4       0     -1      0  ...      20940      19146      19131      2000   
5       0      0      0  ...      19394      19619      20024      2500   
6       0      0      0  ...     542653     483003     473944     55000   
7      -1     -1      0  ...        221       -159        567       380   
8       0      2      0  ...      12211      11793       3719      3329   
9      -2     -2     -2  ...          0      13007      13912         0   
10      0      2      0  ...       2513       1828       3731      2306   
11     -1     -1     -1  ...       8517      22287      13668     21818   
12      0     -1     -1  ...       6500       6500       2870      1000   
13      2      2      0  ...      66782      36137      36894      3200   
14      0      0      0  ...      59696      56875      55512      3000   
15      2      0      0  ...      28771      29531      30211         0   
16      0      2      2  ...      18338      17905      19104      3200   
17      0      0     -1  ...      70074       5856     195599     10358   
18     -2     -2     -2  ...          0          0          0         0   
19     -2     -2     -2  ...          0          0          0         0   

    PAY_AMT2  PAY_AMT3  PAY_AMT4  PAY_AMT5  PAY_AMT6  \
0        689         0         0         0         0   
1       1000      1000      1000         0      2000   
2       1500      1000      1000      1000      5000   
3       2019      1200      1100      1069      1000   
4      36681     10000      9000       689       679   
5       1815       657      1000      1000       800   
6      40000     38000     20239     13750     13770   
7        601         0       581      1687      1542   
8          0       432      1000      1000      1000   
9          0         0     13007      1122         0   
10        12        50       300      3738        66   
11      9966      8583     22301         0      3640   
12      6500      6500      6500      2870         0   
13         0    

In [34]:
# examinar os nomes das colunas
df.columns

Index(['ID', 'LIMIT_BAL', 'SEX', 'EDUCATION', 'MARRIAGE', 'AGE', 'PAY_1',
       'PAY_2', 'PAY_3', 'PAY_4', 'PAY_5', 'PAY_6', 'BILL_AMT1', 'BILL_AMT2',
       'BILL_AMT3', 'BILL_AMT4', 'BILL_AMT5', 'BILL_AMT6', 'PAY_AMT1',
       'PAY_AMT2', 'PAY_AMT3', 'PAY_AMT4', 'PAY_AMT5', 'PAY_AMT6',
       'default payment next month'],
      dtype='object')

Coluna de IDs                     -> ID

Outras colunas                    -> características

Coluna default payment next month -> variável resposta


LIMIT_BAL -> valor do crédito fornecido (em novos dólares taiwaneses (NT)) inclusive o crédito do consumidor individual e familiar (complementar)

SEX -> gênero (1 = masculino; 2 = feminino)

EDUCATION -> instrução (1 = pós-graduação; 2 = universidade; 3 = ensino médio; 4 = outros)

MARRIAGE -> estado civil (1 = casado; 2 = solteiro; 3 = outros)

AGE -> idade

(PAY_1 - PAY_6) -> registros de pagamentos passados (pagamentos mensais passados, registrados de abril a setembro, são armazenados nessas colunas)

PAY_1 -> status reembolso setembro

PAY_2 -> status reembolso agosto

PAY_3 -> status reembolso julho

PAY_4 -> status reembolso junho

PAY_5 -> status reembolso maio

PAY_6 -> status reembolso abril

-- Escala de medida do status reembolso --

-1 = pgto pontual

1 = atraso pgto de um mês

2 = atraso pgto de dois meses

assim por diante até o 9...

9 = atraso pgto de 9 meses ou mais

(BILL_AMT1 - BILL_AMT6) -> valor da fatura (em novos dólares taiwaneses)

BILL_AMT1 -> valor fatura setembro

BILL_AMT2 -> valor fatura agosto

BILL_AMT3 -> valor fatura julho

BILL_AMT4 -> valor fatura junho

BILL_AMT5 -> valor fatura maio

BILL_AMT6 -> valor fatura abril

(PAY_AMT1 - PAY_AMT6) -> valor de pgtos anteriores (em novos dólares taiawneses)

PAY_AMT1 -> valor pago em setembro

PAY_AMT2 -> valor pago em agosto

PAY_AMT3 -> valor pago em julho

PAY_AMT4 -> valor pago em junho


PAY_AMT5 -> valor pago maio

PAY_AMT6 -> valor pago em abril

In [35]:
# observar as primeiras linhas de dados
df.head()

ID  LIMIT_BAL  SEX  EDUCATION  MARRIAGE  AGE PAY_1  PAY_2  \
0  798fc410-45c1      20000    2          2         1   24     2      2   
1  8a8c8f3b-8eb4     120000    2          2         2   26    -1      2   
2  85698822-43f5      90000    2          2         2   34     0      0   
3  0737c11b-be42      50000    2          2         1   37     0      0   
4  3b7f77cc-dbc0      50000    1          2         1   57    -1      0   

   PAY_3  PAY_4  ...  BILL_AMT4  BILL_AMT5  BILL_AMT6  PAY_AMT1  PAY_AMT2  \
0     -1     -1  ...          0          0          0         0       689   
1      0      0  ...       3272       3455       3261         0      1000   
2      0      0  ...      14331      14948      15549      1518      1500   
3      0      0  ...      28314      28959      29547      2000      2019   
4     -1      0  ...      20940      19146      19131      2000     36681   

   PAY_AMT3  PAY_AMT4  PAY_AMT5  PAY_AMT6  default payment next month  
0         0         0         0         0                           1  
1      1000      1000         0      2000                           1  
2      1000      1000      1000      5000                           0  
3      1200      1100      1069      1000                           0  
4     10000      9000       689       679                           0  

[5 rows x 25 columns]

In [36]:
# a coluna ID parece conter valores exclusivos 

# -> verificar se isso ocorre realmente em todo o dataset 

# -> podemos contar o número de valores exclusivos na série ID (primeiro selecionaremos a coluna usando colchetes)

df['ID'].nunique()




29687

In [37]:
# agora vamos verificar o número de linhas do dataset

df.shape

(30000, 25)

In [38]:
# utilizaremos o método .value_counts(), que listará os IDs exclusivos e a frequência com que ocorrem

# primeiro vamos armazenar as contagens de valores em uma variável (id_counts)

id_counts = df['ID'].value_counts()

# exiba os valores armazenados usando o método .head()

id_counts.head()

ad23fe5c-7b09    2
1fb3e3e6-a68d    2
89f8f447-fca8    2
7c9b7473-cc2f    2
90330d02-82d9    2
Name: ID, dtype: int64

In [39]:
# exiba o número de entradas duplicadas agrupadas executando outra contagem de valores

id_counts.value_counts()

1    29374
2      313
Name: ID, dtype: int64

podemos ver que:

    -> 29374 dos IDs ocorrem extamente uma única vez, como esperado
    -> 313 dos Ids no entanto, ocorrem duas vezes (duplicados)
    

**MÁSCARAS BOOLEANAS**

Para ajudar a limpar os dados, introduziremos o conceito de **máscara lógica**, também conhecida como **máscara booleana**.

Para ilustrar como funciona, usaremos **dados sintéticos**, ou seja, dados criados para a exploração ou a ilustração de um conceito

Primeiro, importaremos o pacote **NumPy**, que tem muitos recursos de **geração de números aleatórios**

In [40]:
# importando o pacote NumPy

import numpy as np

In [41]:
# Definindo um seed no gerador de números aleatórios

# se você definir um seed, obterá os mesmos resultados entre as execuções do gerador, 0(zero) -> sempre repetirá o primeiro resultado aleatório
# caso contrário, isso não é garantido

np.random.seed(seed=24)
    

In [42]:
# geraremos 100 inteiros aleatórios, selecionados entre 1 e 5 (inclusive)

#    -> vamos utilizar numpy.random.randint, com os argumentos apropriados

random_integers = np.random.randint(low=1,high=5,size=100)

In [43]:
# examinaremos os cinco primeiros elementos desse array, com random_integers[:5]

random_integers[:5]

array([3, 4, 1, 4, 2])

In [44]:
# suponhamos que quiséssemos conhecer os locais de todos os elementos de random_integers igual a 1

#    -> poderíamos criar uma máscara booleana para fazer isso

is_equal_to_1 = random_integers ==1

In [45]:
# verificando os 5 primeiros elementos

is_equal_to_1[:5]

array([False, False,  True, False, False])

In [46]:
# verificando os 20 primeiros elementos

is_equal_to_1[:20]

array([False, False,  True, False, False, False, False,  True, False,
       False,  True, False, False, False, False, False, False, False,
       False, False])

In [47]:
# agora suponhamos que quiséssemos saber quantos elementos são iguais a 1

#    -> nesse caso, você pode fazer a soma de uma máscara booleana, que interpreta True como 1 e False como 0

In [48]:
# descobrindo quantos elementos são iguais a 1 

sum(is_equal_to_1)

18

Utilizando a **máscara booleana** para **selecionar os elementos do array** 

In [49]:
# indexando arrays utilizando a máscara booleana

random_integers[is_equal_to_1]

array([1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1])

**Máscara booleana** para **localizar IDs duplicados**

In [50]:
# primeiro queremos os índices da série id_counts cuja contagem é 2 -> para localizar as duplicadas

# atribuiremos os Ids duplicados a uma variável chamada dupe_mask

dupe_mask = id_counts ==2

In [51]:
# exibiremos os cinco primeiros Ids duplicados 

dupe_mask[:5]

ad23fe5c-7b09    True
1fb3e3e6-a68d    True
89f8f447-fca8    True
7c9b7473-cc2f    True
90330d02-82d9    True
Name: ID, dtype: bool

In [52]:
# aqui, dupe_mask é a máscara lógica que criamos para armazenar os valores booleanos

Nossa próxima etapa é **usar a máscara lógica para selecionar os Ids duplicados**

In [53]:
# os próprios Ids foram incluídos como índice da série id_count

# podemos acessar o índice para usar nossa máscara lógica para fins de seleção

In [54]:
# acesse o índice de id_count e exiba as cindo primeiras linhas

id_counts.index[:5]

Index(['ad23fe5c-7b09', '1fb3e3e6-a68d', '89f8f447-fca8', '7c9b7473-cc2f',
       '90330d02-82d9'],
      dtype='object')

In [55]:
# selecione e armazene os Ids duplicados em uma nova variável

dupe_ids = id_counts.index[dupe_mask]

In [56]:
# converta dupe_ids em uma lista 

dupe_ids = list(dupe_ids)

In [57]:
# obtenha o tamanho da lista

len(dupe_ids)

313

In [58]:
# altermamos a variável dupe_ids para uma lista

# precisaremos dela nessa forma em etapas futuras

In [59]:
# a lista tem um tamanho igual a 313, que coincide com o número de IDs duplicados

In [60]:
# verificamos os dados da lista dupe_ids, exibindo as cinco primeiras entradas

dupe_ids[0:5]

['ad23fe5c-7b09',
 '1fb3e3e6-a68d',
 '89f8f447-fca8',
 '7c9b7473-cc2f',
 '90330d02-82d9']

In [61]:
# podemos observar que a lista contém as entradas de IDs duplicados

# estamos prontos para examinar os dados na lista dos Ids duplicados

# especificamente, queremos examinar os valores das características:

#    -> para ver o que, se houver algo, há de diferente entre essas entradas duplicadas.

In [62]:
# para isso usaremos os métodos .isin e .loc

In [63]:
# verificamos os três primeiros IDs de nossa lista de duplicadas dupe_ids[0:3]

# primeiro queremos encontrar as linhas que contém esses IDs

# se passarmos essa lista de Ids para o método .isin da série ID, ele criará outra máscara lógica

# que poderemos usar no DataFrame maior para exibir as linhas que contém os IDs

# o.isin ficará aninhado em uma instrução .loc

# .loc indexa o DataFrame para a seleção local de todas as linhas que contém "True" na máscara lógica

# o segundo argumento da instrução de indexação .loc é : , que implica que todas as colunas serão seleciondas



In [64]:
# ao executar as próximas etapas estaremos basicamente filtrando o DataFrame para visualizar todas as colunas dos três primeiros IDs duplicados.

# examine os dados de IDs duplicados

df.loc[df['ID'].isin(dupe_ids[0:3]),:].head(10)

ID  LIMIT_BAL  SEX  EDUCATION  MARRIAGE  AGE PAY_1  PAY_2  \
5033   89f8f447-fca8     320000    2          2         1   32     0      0   
5133   89f8f447-fca8          0    0          0         0    0     0      0   
16727  1fb3e3e6-a68d      80000    1          2         2   33     2      2   
16827  1fb3e3e6-a68d          0    0          0         0    0     0      0   
29685  ad23fe5c-7b09      50000    1          3         1   32     0      0   
29785  ad23fe5c-7b09          0    0          0         0    0     0      0   

       PAY_3  PAY_4  ...  BILL_AMT4  BILL_AMT5  BILL_AMT6  PAY_AMT1  PAY_AMT2  \
5033       0      0  ...     169371     172868     150827      8000      8000   
5133       0      0  ...          0          0          0         0         0   
16727      0      0  ...      27394      29922      31879         0      2000   
16827      0      0  ...          0          0          0         0         0   
29685      0      0  ...      12882       8131       3983      3000      2871   
29785      0      0  ...          0          0          0         0         0   

       PAY_AMT3  PAY_AMT4  PAY_AMT5  PAY_AMT6  default payment next month  
5033       5500      6100      6000      5000                           0  
5133          0         0         0         0                           0  
16727      2000      3000      2600         0                           1  
16827         0         0         0         0                           0  
29685      1000       163      3983      3771                           1  
29785         0         0         0         0                           0  

[6 rows x 25 columns]

In [66]:
# estamos vendo que cada ID duplicado parece ter uma linha de dados válidos e outra somente com zeros

# o mais fácil a se fazer é excluir as linhas que somente contém zeros

# primeiro vamos encontrar as linhas que só tem zeros (exceto na primeira coluna que contém os IDs)

# podemos encontrar as entradas do DataFrame que são iguais a zero

# criando uma matriz booleana com o mesmo tamanho do DataFrame, usando a condição "é igual a zero"



In [67]:
# criando uma matriz booleana com o mesmo tamanho do DataFrame

df_zero_mask = df == 0

In [69]:
# nas próximas etapas utilizaremos o df_zero_mask, que é outro DataFrame, contendo valores booleanos


O objetivo é criarmos uma série booleana, **feature_zero_mask**, que identifique cada linha em que todos os elementos 
a partir da segunda coluna (as características e a resposta, mas não os IDs), sejam 0. 

In [70]:
# para isso, primeiro temos de indexar df_zero_mask usando o método de indexação de inteiros (.iloc)

# nesse método, passaremos (:) para examinar todas as linhas

# (1:) para examinar todas as colunas a partir da segunda (índice 1)

# para concluir, aplicaremos o método all() ao longo do eixo da coluna (axis=1)

# e ele retornará True somente se todas as colunas dessa linha forem iguais a True. 

In [71]:
# criando a série booleana feature_zero_mask

feature_zero_mask = df_zero_mask.iloc[:,1:].all(axis=1)

In [72]:
# calculando a soma da série booleana

sum(feature_zero_mask)

315

In [74]:
# esse é o número de linhas somente com zeros, exceto pelo ID

# a saída anterior nos diz que 315 linhas têm zeros para cada coluna, exceto a primeira

# esse número é maior do que o número de IDs duplicados (313), logo;

# se excluirmos todas as "linhas de zeros", podemos nos livrar do problema dos Ids duplicados

In [75]:
# limpando o DataFrame

df_clean_1 = df.loc[~feature_zero_mask,:].copy()

In [76]:
# oberve que usamos o método.copy() após a operação de indexação .loc 

# para criar uma cópia dessa saída, em vez de visualizarmos o DataFrame original

# dentro do método .loc, usamos o operador lógico not ~ 

# para selecionar todas as linhas que não têm zeros

# para todas as características e a resposta, e:

# para selecionar todas as colunas


Agora queremos saber se o **número de linhas restantes** é igual ao número de **IDs exclusivos**

In [77]:
# verificando o número de linhas e colunas do DataFrame limpo com todos os dados

df_clean_1.shape

(29685, 25)

In [78]:
# verificando o número de IDs exclusivos do DataFrame limpo

df_clean_1['ID'].nunique()

29685

In [79]:
# podemos ver que eliminimos com sucesso os IDs duplicados

# já que o número de IDs exclusivos é igual ao número de linhas

In [81]:
# ---- Essa foi uma introdução a algumas técnicas do pandas para indexação e caracterização de dados ---- 
#      (indexação lógica e pandas)

# ---- Esse foi um probema básico de qualidade de dados -> quais amostras estavam presentes, com base nos metadados

# ---- Extraídos os IDs duplicados, estamos prontos para examinar os dados propriamente ditos ----

# ---- As características e a resposta ----

Examinar os valores das **características e da resposta**

-> os dados que usaremos para **desenvolver nosso modelo preditivo**